### PR TITLE
Use asynchronous exec to eliminate freeze on save

### PR DESCRIPTION
**Pull request by 1lann**
Currently if you're using the goreturns or goimports format tools, Atom can freeze for significant amounts of time. For one of my projects, saving completely freezes Atom for ~600 ms, which makes Atom feel slow.

This is due to the fact that the format tools are executed by synchronous operations, which this pull request has changed to be asynchronous so that Atom doesn't freeze while saving. The only downside to this is that any edits made in this short time frame while the formatter is being run will be lost, but I think it's worth the sacrifice to make Atom feel smoother.

### CODE REVIEW

**Comment by 1lann** [comment]
Unfortunately this breaks the tests which assume saving is synchronous 😞 , I'm not sure how to fix those as I don't develop much in JavaScript or work with Atom. Sorry about that.

**Comment by joefitzgerald** [comment]
This is the way I initially implemented. I had to revert to being synchronous due to the save API in atom being synchronous. Without this, there is a flicker in the buffer as the entire buffer is re-tokenized via the grammar. The solution is to make the Atom Save API asynchronous.

**Comment by 1lann** [comment]
Hm, there doesn't seem to be a flicker when I've used it asynchronously, at least not anymore (probably due to the use of setTextViaDiff, or were you using that before too?).

**Comment by 1lann** [comment]
I just tested it out. A flicker occurs if you use setText, but not if you use setTextViaDiff.

**Comment by joefitzgerald** [comment]
Good point; I wasn't using it before, and that would probably improve things. The other issue is there is no way to save the resulting buffer changes without finding yourself in the beginnings of a loop. You could theoretically guard against infinite recursion but then you're unnecessarily running linters, and gofmt / goimports twice.

**Comment by joefitzgerald** [comment]
Oh and because goimports makes an otherwise invalid file valid, you'll end up with spurious lint errors until you re-save the file.

**Comment by 1lann** [comment]
> You could theoretically guard against infinite recursion but then you're unnecessarily running linters, and gofmt / goimports twice.

My current changes should guard against infinite recursion (unless some other package modifies the file again on save). I'm aware that it is unnecessarily running other linters twice, but at least in my opinion, I'd rather have a editor that doesn't freeze but does more work in the background, than one that freezes for half a second whenever I hit save.

**Comment by 1lann** [comment]
Would you reckon having a configurable option for the user to choose between saving synchronously or asynchronously would be better?

**Comment by abijr** [comment]
Any updates on this? Sometimes I get a freeze of more than just a few seconds, and it is incredibly annoying.

**Comment by joefitzgerald** [comment]
https://groups.google.com/forum/#!topic/golang-codereviews/P83_TTz3fUc

TL;DR: _Update goimports_ (`go get -u golang.org/x/tools/cmd/goimports`)

**Comment by abijr** [comment]
thanks, that solved it for me

**Comment by 1lann** [comment]
Awesome. Looks like the new goimports is much more faster now, making this pull request obsolete.
